Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1338 @@
+
+The solution here is simple enough, just use the `%shift` directive:
+
+```diff
+-  : if exp then exp          { EIfThen (L.rtRange $1 <-> info $4) $2 $4 }
++  : if exp then exp %shift   { EIfThen (L.rtRange $1 <-> info $4) $2 $4 }
+   | if exp then exp else exp { EIfThenElse (L.rtRange $1 <-> info $6) $2 $4 $6 }
+```
+
+As a sanity check in GHCi, we can check the (prettified and simplified) output for the ambiguous expression:
+
+```haskell
+>>> runAlex "let x = if a then if b then c else d" parseMiniML
+Right
+  [ Dec _ (Name _ "x") [] Nothing           -- let x =
+    (EIfThen _ (EVar _ (Name _ "a"))        --   if a then
+      (EIfThenElse _ (EVar _ (Name _ "b"))  --     if b
+      (EVar _ (Name _ "c"))                 --     then c
+      (EVar _ (Name _ "d"))))               --     else d
+  ]
+```
+
+You can find the source code up to this point [here](https://gist.github.com/heitor-lassarote/ebfddb9b354bb63dda2f952947d09f84).
+
+### Parsing negations
+
+Now, let's make it, so we are able to negate expressions.
+This case is pretty simple, just stick this definition in `exp`:
+
+```happy
+  | '-' exp                  { ENeg (L.rtRange $1 <-> info $2) $2 }
+```
+
+This will also allow us to parse interesting cases, such as `-if True then 1 else 2` or `-succ 2`.
+Try it in GHCi, both cases are valid Haskell code, and we also make it valid MiniML code now.
+
+#### Parsing binary operators
+
+Now let's move on with parsing binary operators, where Happy will make us unhappy again.
+
+First, add new productions to parse each existing operator to `exp`:
+
+```happy
+  -- Arithmetic operators
+  | exp '+'  exp             { EBinOp (info $1 <-> info $3) $1 (Plus (L.rtRange $2)) $3 }
+  | exp '-'  exp             { EBinOp (info $1 <-> info $3) $1 (Minus (L.rtRange $2)) $3 }
+  | exp '*'  exp             { EBinOp (info $1 <-> info $3) $1 (Times (L.rtRange $2)) $3 }
+  | exp '/'  exp             { EBinOp (info $1 <-> info $3) $1 (Divide (L.rtRange $2)) $3 }
+  -- Comparison operators
+  | exp '='  exp             { EBinOp (info $1 <-> info $3) $1 (Eq (L.rtRange $2)) $3 }
+  | exp '<>' exp             { EBinOp (info $1 <-> info $3) $1 (Neq (L.rtRange $2)) $3 }
+  | exp '<'  exp             { EBinOp (info $1 <-> info $3) $1 (Lt (L.rtRange $2)) $3 }
+  | exp '<=' exp             { EBinOp (info $1 <-> info $3) $1 (Le (L.rtRange $2)) $3 }
+  | exp '>'  exp             { EBinOp (info $1 <-> info $3) $1 (Gt (L.rtRange $2)) $3 }
+  | exp '>=' exp             { EBinOp (info $1 <-> info $3) $1 (Ge (L.rtRange $2)) $3 }
+  -- Logical operators
+  | exp '&'  exp             { EBinOp (info $1 <-> info $3) $1 (And (L.rtRange $2)) $3 }
+  | exp '|'  exp             { EBinOp (info $1 <-> info $3) $1 (Or (L.rtRange $2)) $3 }
+```
+
+n.b.: Do NOT extract these operators to a new production and change it to `exp operator exp` because of a limitation in Happy. This will cause conflicts that will be pretty difficult to resolve.
```

Comment:
BTW, you can actually work around this by using `%prec`, but you _will_ have to list all lookahead tokens in the operator precedence table.

Reply:
I mentioned what you said in the article.